### ISSUE TITLE
Crash to desktop when loading priceguide data version 2021.1.4

### ISSUE THREAD

**Issue by chiminirc** [rgriebl/brickstore#86]
Thank you for the new build.  I saw notes about fixes for a possible bug with identical part/color lots.
Windows 10, 64 bit verison of 2021.1.4

Testing Ninjago City Docks and it always crashes.
Kit loaded via CTRL i + CTRL i
typed ninjago city and selected kit .Import was quick.  
CTRL G and accepted defaults "last 6 month sales" and "Average"
"Verifying price guide data" progress bar shows.  Before it finishes, the dialog "prices of the selected items  have been updated..." pops over.
Click that and the window disappears, you see the incomplete progress bar again and then crash to desktop. 

I thought maybe the fact the images didn't download matter.  I scrolled through the document to get all images cached.  Same behavior and crash.

I then marked all items "included" and consolidated so all lots are unique. Same behavior and crash.

What I can say is that is appears to keep loading more and more data until eventually after 4-5 interactions, it doesn't crash. 

It is almost as if the finished dialog has a timeout that will pop over the progress regardless if it is done or not. I can also say that if I load 100 rows at a time it won't crash until I eventually get through the entire document.

**Comment by paramecie** [comment]
I'm testing this for now it has no problem:
![image](https://user-images.githubusercontent.com/32078594/104764860-43c2fe80-5768-11eb-86cd-47930c7e73e4.png)

Zero crash (2021.1.4 on Win 7/64)

I can see no "Verifying price guide data" progress bar ???

Which Currency are you using? Here the test was in USD.

**Comment by chiminirc** [comment]
USD

**Comment by paramecie** [comment]
Then no problem here: Import Inventory is ultra fast, CTRL+A CTRL+G took maybe 1 minute, then maybe 1 extra minute to download pictures.

May be related to high trafic on BrickLink? I'll test more.

**Comment by chiminirc** [comment]
In order to replicate it you must be loading a kit that isn't already in the priceguide cache. If you load a kit with most data downloaded, it doesn't crash.

Can you try renaming your brickstore priceguide cache, relaunch, and try pulling down a big set like ninjago city docks?  I can confirm I can immediately do this with brickstock after the crash with a forced price guide download. All data pulled down in 3 seconds. Doesn't feel like high traffic.

**Comment by paramecie** [comment]
I may have tested part out for a couple of sets only, and not this one.
I don't use BrickStock right now - just for testing, not prod.

But I'll test more...

**Comment by chiminirc** [comment]
I'm trying not to use brickstock but still have it loaded to verify things like this.  i also need to keep it around for pricing.  The new brickstore is superior for parting out sets now that I can easily break things like minifigs in parts without forcing a new document creation. I just need the stability in brickstore to fully retire brickstock.

**Comment by paramecie** [comment]
>  I just need the stability in brickstore to fully retire brickstock.

Oh me too, me too - I'd be soooo happy!

**Comment by paramecie** [comment]
> In order to replicate it you must be loading a kit that isn't already in the priceguide cache. If you load a kit with most data downloaded, it doesn't crash.

I did what you said: I deleted the whole cache, including DB, prices and pictures.

![image](https://user-images.githubusercontent.com/32078594/104774902-c0110e00-5777-11eb-945d-9fd1514890d7.png)

I then opened BrickStore, updated DB, CTRL+I CTRL+I downloaded the 736 lots of the Ninjago City Docks with zero problem.
CTRL A CTRL G for setting prices to Avg (while updating the pictures), no problem.

The whole took less than 2 minutes, from scratch.

Note: 734 lots twice is already 1500+ requests to BrickLink, don't forget the 5000 daily limit...

**Comment by paramecie** [comment]
Still crashing?

> "Verifying price guide data" progress bar shows.

I've never seen a "progress bar" showing "Verifying price guide data".
It's in the translation file, so you obviously didn't invent it ;-)

Maybe because it's too fast to be seen?
Mmmmmh maybe your C: drive is an HDD (here SSD)?

Is your C: filesystem has room and isn't corrupted - verify using chkdsk or right click on the disk, tools, verify drive (Win 7 anyway, don't remember on Win 10)?

**Comment by rgriebl** [comment]
Now that is weird. You seem to have slow PC (the progress dialog pop only pops up, if an action takes longer than ~1.5 seconds), but you must be sitting right next to the BL servers: you are receiving the PG downloads faster than your PC can send out new requests.
This is something I never thought of and I can't reproduce it even on my 600MBit line. I need a bit to think about how to redesign this part of the code.
Maybe I've just sped up the downloading too much with my optimizations in the .4 release, because the algorithm behind the set-to-priceguide function is still the same as in the old BrickStore/-Stock versions. ;)

**Comment by chiminirc** [comment]
core i7 8850h processor
16gb ram.
500gb SSD with plenty of space free.
I actually have crappy internet. 15-50mbit

**Comment by chiminirc** [comment]
Regarding the API limit.  I'm quite confident I am not hitting that limit.  It has happened before and it's usually quick to see by visiting bricklink and not seeing prices.  I can also switch IP via a different ISP or VPN tunnel. I can also come right behind this crash with brickstock and pull the same kit without error (via forced downloads so I know I'm pulling every lot instead of a mix of cache data and API data)

This is a video capture of the crash. 
https://www.youtube.com/watch?v=E30T7JCITYE

**Comment by chiminirc** [comment]
In the video. The first is the large ninjago set which I had already managed to pull all the lots down so it was all on local storage.  Very quick to load all the values.  After that I did Green Grocer. that is a large set I know I haven't cached price data on.  This replicates the bug. Screenshot of priceguide configs below.

Does it makes sense to retest with any of the following?
-wipe the local cache of all priceguide txt files?
-run the app limited to a single core (unless multithreading is tripping things up? waiting on branched thread?)
-Try the 32bit version?

![image](https://user-images.githubusercontent.com/76887746/104796161-6dd0ec80-5779-11eb-9101-94e002595b2c.png)

**Comment by rgriebl** [comment]
Could you please try this version that's currently being built? Doesn't matter if 32 or 64 bits:

https://github.com/rgriebl/brickstore/actions/runs/489478507

I think I have a fix, but I just can't trigger the bug to verify it here.

**Comment by paramecie** [comment]
#88 has been solved by the above fix!
This current one may be fixed too?

Robert is slow, but he's fast when he isn't slow :O)

**Comment by chiminirc** [comment]
I just loaded it up now.  It seems to have fixed it.
It is worth mentioning my first test was Green Grocer which I had already pulled down into the cache.  The progress bar ran for about 2-3 seconds instead of being near instant like the prior version..  I then followed up with FERRIS WHEEL and HAUNTED HOUSE, two large sets.  Took about 3-5 seconds to pull down all the data.  In all cases the completion dialog didn't pop over the progress bar.  The progress bar ran, then was replaced with the completion window.
